cmd/pj-rehearse-metrics/pj-rehearse-metrics.go

```diff
@@ -0,0 +1,185 @@
+		return
+	}
+
+	if err := os.MkdirAll(prDir, 0755); err != nil {
+		fmt.Printf("could not create a directory for PR err=%v\n", err)
+		return
+	}
+
+	if err := ioutil.WriteFile(localMetricsPath, data, 0644); err != nil {
+		fmt.Printf("could not write metrics to %s err=%v\n", localMetricsPath, err)
+		return
+	}
+}
+
+func (s *scraper) scrapePr(prefix string) {
+	prNumber := path.Base(prefix)
+	jobPrefix := path.Join(prefix, rehearseJobName)
+	jobPrefix = fmt.Sprintf("%s/", jobPrefix) // Needed because path.Join() removes trailing slash
+
+	runs := s.bucket.Objects(s.ctx, &storage.Query{Prefix: jobPrefix, Delimiter: "/"})
+	for {
+		attrs, err := runs.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			fmt.Printf("Next() failed err=%v\n", err)
+			break
+		}
+
+		// If we see non-empty `Name`, we got a full object name = a file
+		// At this level, we only care about dirs, not files.
+		if attrs.Name != "" {
+			continue
+		}
+
+		s.scrapeRun(attrs.Prefix, prNumber)
+	}
+}
+
+type options struct {
+	cacheDir string
+}
+
+func gatherOptions() options {
+	o := options{}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+
+	fs.StringVar(&o.cacheDir, "cache-dir", "", "Path to a directory where scraped metrics data will be saved. If not provided, cache will not be used.")
+
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		logrus.WithError(err).Fatal("Failed to parse options")
+	}
+	return o
+}
+
+func main() {
+	o := gatherOptions()
+	var err error
+
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to create client")
+	}
+
+	if o.cacheDir == "" {
+		o.cacheDir, err = ioutil.TempDir("", "")
+		if err != nil {
+			logrus.WithError(err).Fatal("Failed to create a temporary directory")
+		}
+		defer func() {
+			if err := os.RemoveAll(o.cacheDir); err != nil {
+				logrus.Warnf("Failed to clean up a temporary directory %s (err=%v)", o.cacheDir, err)
+			}
+		}()
+	}
+
+	bucket := client.Bucket(originCiBucket)
+	pullRequests := bucket.Objects(ctx, &storage.Query{
+		Prefix:    "pr-logs/pull/openshift_release/",
+		Delimiter: "/",
+	})
+
+	scraper := scraper{ctx: ctx, bucket: bucket, cacheDir: o.cacheDir}
+	prPrefixes := make(chan string, 100)
+	progress := make(chan string, 100)
+
+	sem := semaphore.NewWeighted(workersCount)
+
+	for w := 0; w < workersCount; w++ {
+		if err := sem.Acquire(ctx, 1); err != nil {
+			logrus.WithError(err).Error("Failed to acquire semaphore to spawn a worker")
+			break
+		}
+		go func() {
+			defer sem.Release(1)
+			for prefix := range prPrefixes {
+				scraper.scrapePr(prefix)
+				progress <- prefix
+			}
+		}()
+	}
+
+	go func() {
+		defer close(progress)
+		for {
+			attrs, err := pullRequests.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				fmt.Printf("Next() failed: err=%v\n", err)
+				break
+			}
+
+			if attrs.Name != "" {
+				continue
+			}
+
+			prPrefixes <- attrs.Prefix
```

Comment:
You could also just remove the need for `prPrefixes`, run the content of this goroutine in the main thread and attempt to spawn a worker here -- we would read new prefixes as we need to instead of upfront but that's not a huge hit in cost, right?

Reply:
You mean to spawn short-lived goroutines to scrape just a single PR? It's possible, but I'm not sure it's better than the long-lived workers over the channel (not worse either, just different). 

I do not understand the comment about reading prefixes upfront, though - reading prefixes would still need to be implemented via looping over the `pullRequests` iterator, no? This goroutine reads prefixes and sends them to the channel as fast as the workers pop them, where's the upfront read?

Reply:
The approach right now reads them all upfront and sends to the channel, the channel holds all of the values (sans whichever have started to be processed). Doing it just-in-time will read as we process.